Sentiment analysis

femaleprog opened the pull request:
### Pull Request: Improvements for Sentiment Analysis (Step 1)

Hello reviewers,

In this pull request, I'm submitting the code changes that correspond to Step 1 of our project plan. The focus of this step was to improve the sentiment analysis code and prepare it for further experimentation. Here's a summary of the changes made in this step:

- All code updates related to sentiment analysis are included in this commit.
- Code formatting, spellings, and basic code standards have been addressed to ensure consistency.
- The code has been organized and documented for easier understanding.

The purpose of this pull request is to mark the completion of Step 1. In the upcoming steps, I plan to perform the following tasks:

1. Calculate accuracy, precision, and recall: I will design an experiment to evaluate these metrics for the sentiment analysis task.
2. Design an experiment: The focus will be on evaluating the accuracy, precision, and recall of the sentiment analysis model.
3. Explore causal coherence: Once the metrics are evaluated, I will delve into identifying types of coherence, starting with causal coherence.

Your feedback on the current changes and future plans is greatly appreciated. Feel free to provide any insights or suggestions that can guide our next steps. Thank you for your time and collaboration!

Best regards,
Soukaina

femaleprog committed:
Initial commit

femaleprog committed:
seconde commit : cleaner code

femaleprog committed:
specificity metrics and new data

smirik committed:
Create python file for the review purposes

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+
+
+# %%
+data = pd.read_csv('/kaggle/input/dataset0/data.csv')
```
Please, use local file path.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# ## 1. Specificity
+
+# %%
+import nltk
```
Move all imports to the top.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+from nltk.chunk import ne_chunk
+
+# Tokenization
+def tokenize_text(text):
```
Python supports type hints. It might be a good idea to use them. I.e., for this example:

`def tokenize_text(text: str) -> str:`

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+
+# Tokenization
+def tokenize_text(text):
+    return sent_tokenize(text)
```
Note that this function is a little bit useless because it just forwards data to another function. The only value might be if you expect to use another function, for some reason.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# Part-of-Speech (POS) Tagging
+def pos_tagging(text):
+    tokens = word_tokenize(text)
+    return pos_tag(tokens)
```
You don't need two lines for this code.

```suggestion
    return pos_tag(word_tokenize(text))
```

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+def named_entity_recognition(text):
+    tokens = word_tokenize(text)
+    tagged = pos_tag(tokens)
+    ne_tree = ne_chunk(tagged)
```
Add an empty line to visualise the logical code blocks.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# Event Extraction
+def extract_events(text):
+    tokens = word_tokenize(text)
+    tagged = pos_tag(tokens)
```
Add an empty line.

smirik replied:
Also, you don't line `tokens` variable. Hence, you can compile both lines into one.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+    events = []
+    current_event = []
+    for tag in tagged:
+        if tag[1].startswith('VB'):
```
What does it mean 'VB'?

smirik replied:
It is a good approach when your code is self-explonatory. This line is not intuitive. Hence, you might want to improve it. As an option, you can create a function that will perform this check. While it might look useless (this function invokes another; thus, my comment above is valid for this case too), when you name properly this function, it will make the code readable.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+    for tag in tagged:
+        if tag[1].startswith('VB'):
+            current_event.append(tag[0])
+        elif current_event:
```
I'm not sure, but this condition seems to be true all the time because `current_event` is a list. Moreover, `if` and `elif` statement should contain a boolean condition, whereas `current_event` is a list.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+        elif current_event:
+            events.append(' '.join(current_event))
+            current_event = []
+    if current_event:
```
See comment above.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+print(data.iloc[0]['Story'])
+
+# %%
+def specificity_ne(story):
```
It's a bad behaviour to print anything within a function. Return all the data you need and then print the output.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# Testing
+
+# %%
+specificity_ne(data.iloc[0]['Story'])
```
Usually, functions should return something. Yes, you may create functions returning void, but usually, this is a bad practice.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+print('Named Entities:', named_entities)
+
+# %%
+def specificity_ne_spacy(story):
```
See comments above.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# # 2.Personal Context 
+
+# %%
+import nltk
```
Move imports to the top. Combine similar imports.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# Understanding the significance of individual words and their weights within a topic can help provide insights into the key themes and subjects present in the text.
+
+# %%
+def test_personal_context(text):
```
See comments above.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+test_personal_context(data.iloc[0]['Story'])
+
+# %%
+test_personal_context("i lost my best friend and im so happy")
```
Good move to test some obvious examples where we have an expected return. However, it would be nice to create some of them, store the expected result within a csv file, and compare within a script actual and expected results. You can use `pytest` as a framework for such a task.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# ## 3.Sensory details 
+
+# %%
+import nltk
```
Move to the top.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+def analyze_sensory_details(text):
+    sentences = nltk.sent_tokenize(text)
+
+    sensory_keywords = {
```
This looks like a global declaration of sensory keywords.

Move this declaration to the const on the top level of a script. Or create a top-level variable and use it as an argument of a function. Another (better) option: move it to some config file in the proper format to support easy modifications.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    sensory_details = []
+    for sentence in sentences:
```
Usually, it is a bad practice when your nesting level is greater than 4.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+
+    sensory_details = []
+    for sentence in sentences:
+        lower_sentence = sentence.lower()
```
Also, I'm not sure whether you are accurate here because you try to find the exact match, whereas in reality, you might have different forms of verbs (i.e., felt instead of feel, etc). You can adjust your code if you find more general solution.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+        for sense, keywords in sensory_keywords.items():
+            for keyword in keywords:
+                if keyword in lower_sentence:
+                    sensory_details.append({'sense': sense, 'sentence': sentence})
```
You use 'greedy' algorithm approach, but what if there are two different keywords in the list?

femaleprog replied:
I might be wrong but I think that since our primary goal is to detect whether a sensory keyword appears in a sentence in this code design, a greedy approach simplifies the task.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+
+
+# %%
+def test_sensory_details(text):
```
It's not about testing, it's about printing. Hence, the right name should be `print_sensory_details`. Also, see my comments above.

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+# using emotional analysis 
+
+# %%
+from nltk.sentiment import SentimentIntensityAnalyzer
```
Move to the top

smirik commented on coherence-pr-review.py:
```diff
@@ -0,0 +1,266 @@
+
+speech = "The loss of my father will forever leave an indelible mark on my heart. But im so happy that he died"
+
+emotion_scores = analyze_emotional_tone(speech)
```
You get the raw value of scores. But how to convert it to the result? What are the thresholds? I'd encourage you to create another function that will make a decision based on the value of emotional scores.

femaleprog replied:
Yes, creating thresholds is one of the things I did in the latest version of the notebook. You can find it in the "metrics" branch.

smirik reviewed (changes requested):
Hello there!

It's a little bit tricky to code review Jupyter's notebooks because they have no native support on GitHub (through the review process). Hence, it might be a good idea to move the code that is 'general' to some python files inside of 'src' or coherence-in-ml subdirectory.

**Comments:**

1. Let's rename 'Data' folder to 'data'. Usually, developers prefer to use lowercase unless another is specified in the framework or tool you are using. It helps to prevent some errors across multiple systems.
2. Let's rename data.csv to 'stories.csv' to make it more convenient.
3. Note that for both tasks above, use `git mv` command (rename/move through git) to avoid unnecessary deletion and addition.
4. See the comments to the python file. You can continue with Jupyter Notebook & apply changes there. Then you can remove all the comments by the end of this exercise. Otherwise, you can move the universal code (i.e., functions) to a python file, apply changes there, and modify Jupyter notebook to import these functions and use them there.

femaleprog committed:
Delete coherence-in-ml.ipynb

Remove Jupyter Notebook, transition to Python files

In preparation for transitioning from Jupyter Notebook to Python files for improved code organization and maintainability, this commit removes the Jupyter Notebook "coherence-in-ml.ipynb." The code and content from the notebook will be migrated to separate Python files

femaleprog committed:
add seperate python files for specificity, personal context and sensory details analysis

femaleprog committed:
changes

femaleprog committed:
Move files to root of sentiment-analysis branch

femaleprog committed:
Add new seperate python files

femaleprog commented:
Hi @smirik,

I've made some changes in this branch since our last discussion. I wanted to bring your attention to these updates and kindly request your review. Please take a look and let me know your thoughts.

Thank you!
